### ISSUE TITLE
Injected IEnumerables dependencies are singleton and not new instances

### ISSUE THREAD

**Issue by messengert** [seesharper/LightInject#307]
I have classes that load specific file types based on file extension. An  example would be:

     // Constructor...
     public ArSumFileLoader (
            IPostDate postDate,
            ILogger logger,
            IDbMainRepository dbMainRepo,
            IDataTracker codeTracker )
            : base( logger, dbMainRepo )
        {
            _postDate = postDate;
            _codeTracker = codeTracker;
        }

       //
      public Process( filepath )
      {
          ...
      }

and registered as:

     // service name is the file extension
     container.Register( typeof( IFileLoader ), type, serviceName );

I also have a CodeTracking class that each IFileLoader is dependant on and is supposed be a NEW instance. They are registered as such:

     container.Register<IDataTracker, CodeTracker>( );

I then have a class that is injected at the contructor that brings in all the IFileLoaders, among other things. The "IEnumerable<IFileLoader> fileLoaders" are converted to a dictionary for quick look up based on the file extension.

        public ClientImporter (
            ILogger logger,
            IDbStagingRepository dbStagingRepo,
            IDbMainRepository dbMainRepo,
            IDbCommonRepository dbCommonRepo,
            IEnumerable<IEtlProcessor> processors,
            IEnumerable<IFileLoader> fileLoaders )
        {
            _logger = logger;
            _dbStagingRepo = dbStagingRepo;
            _dbMainRepo = dbMainRepo;
            _dbCommonRepo = dbCommonRepo;

            _processors = processors.ToDictionary( k => k.ServiceName, v => v );
            _loaders = fileLoaders.ToDictionary( k => k.ServiceName, v => v );
        }

        public IEnumerable<string> ProcessFile( string filepatth )
        {
           var extension = Path.GetExtension( filepath );
           var loader = _loaders[ extension ];
           loader.Process( filepath );

           return loader.Tracker.GetTackedClientCodes( );
        }

At first glance every seems to be fine, All parameters are injected as I would expect them to be, but all is not well. So, to recap... the ClientImporter is injected with an Enumeration of IFileLoaders, and each loader is injected with an IDataTracker object. The problem is that after a loader is done processing the file, ALL loaders in the dictionary have the same values in each loaders IDataTracker object. This should not be the case.

Because of the way everything is registered, each IFileLoader object should have its OWN...NEW instance of a IDataTracker object, but it is acting like the IDataTracker object was registered as an instance. 

I am on LightInject version: 4.0.11.0
I am on Windows 10

Any help would be greatly appreciated.
Thomas

**Comment by seesharper** [comment]
Could you produce a fully working example of what you are trying to do? 

The easy way is to push a repo up on GitHub that we can PR into.

**Comment by messengert** [comment]
In the process of putting together a small replica sample for you I discovered that I had accidentally registered the IDataTracker object twice. One as an instance and then later as a transient. Once I spotted and removed the instance registration the code works as expected. I found it interesting the I could have it registered the object twice in two different styles 'instance' and 'transient' and container was none the wiser. I would have thought the latter transient registration would have overwritten the instance registration. Hmmm.

So, for now...all is right with the universe again. :)
